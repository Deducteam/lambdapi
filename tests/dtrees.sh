#!/bin/sh
# Test Lambdapi decision-tree command

set -uf

ko() {
    printf '\033[31mKO\033[0m %s\n' "$1"
    exit 1
}

ok() {
    printf '\033[32mOK\033[0m %s\n' "$1"
}

LAMBDAPI="lambdapi decision-tree --verbose 0 \
--map-dir=tests:tests"

out="$(${LAMBDAPI} tests.OK.nat.+)"
if [ -z "$out" ]; then
    ko 'tests.OK.nat.+'
else
    ok 'tests.OK.nat.+'
fi

# Escaped characters (with no rule)
out="$(${LAMBDAPI} 'tests.OK.Escaped.{|KIND|}' 2>/dev/null)"
if [ "$?" = 1 ]; then
    ko 'tests.OK.Escaped.{|KIND|}'
fi
if [ -z "$out" ]; then
    ok 'tests.OK.Escaped.{|KIND|}'
fi

# Escaped with dots (no rule on symbol)
out="$(${LAMBDAPI} 'tests.OK.Escaped.{|KIND.OF.BLUE|}' 2>/dev/null)"
if [ "$?" = 1 ]; then
    ko 'tests.OK.Escaped.{|KIND.OF.BLUE|}'
fi
if [ -z "$out" ]; then
    ok 'tests.OK.Escaped.{|KIND.OF.BLUE|}'
fi

# Escaped with rules
out="$(${LAMBDAPI} 'tests.OK.Escaped.{|set|}' 2>/dev/null)"
if [ -z "$out" ]; then
    ko 'tests.OK.Escaped.{|set|}'
else
    ok 'tests.OK.Escaped.{|set|}'
fi

# Ghost symbols
# TODO put back tests.OK.unif_hint.#equiv when OK/unif_hint.lp is fixed
# out="$(${LAMBDAPI} 'tests.OK.unif_hint.#equiv' 2>/dev/null)"
# if [ "$?" = 1 ] || [ -z "$out" ]; then
#     ko 'tests.OK.unif_hint.#equiv'
# else
#     ok 'tests.OK.unif_hint.#equiv'
# fi
exit 0
