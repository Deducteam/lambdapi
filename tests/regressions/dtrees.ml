(** Test decision trees to graph export *)

open Common
open Handle
open Core

let () =
  Library.set_lib_root (Some "/tmp");
  Timed.(Console.verbose := 0);
  let sign = Compile.PureUpToSign.compile_file "../OK/boolean.lp" in
  let ss = Sig_state.of_sign sign in
  (* Regular symbol *)
  let sym =
    Sig_state.find_sym ~prt:true ~prv:true ss
      (Pos.none ([ "tests"; "OK"; "boolean" ], "bool_or"))
  in
  Format.printf "=> tests.OK.boolean.bool_or@\n";
  Tool.Tree_graphviz.to_dot Format.std_formatter sym;
  (* Ghost symbol *)
  (* We don't use [Pure] to keep rules added to unif hint symbols *)
  ignore @@ Compile.compile_file "../OK/unif_hint.lp";
  Format.printf "=> ghost symbol â‰¡ from tests.OK.unif_hint@\n";
  Tool.Tree_graphviz.to_dot Format.std_formatter Unif_rule.equiv
