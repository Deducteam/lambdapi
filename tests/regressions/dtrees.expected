=> tests.OK.boolean.bool_or
graph {
       0 [label="0" shape="box"];
       1 [label="0" shape="box"];
       0 -- 1 [label=<false<sub>0</sub>>];
       2 [label="$b"];
       1 -- 2 [label=<*>];
       3 [label="0" shape="box"];
       0 -- 3 [label=<true<sub>0</sub>>];
       4 [label="true"];
       3 -- 4 [label=<*>];
       5 [label="0"];
       0 -- 5 [label=<*>];
       6 [label="$b"];
       5 -- 6 [label=<false<sub>0</sub>>];
       7 [label="true"];
       5 -- 7 [label=<true<sub>0</sub>>];
}
=> ghost symbol ≡ from tests.OK.unif_hint
graph {
       0 [label="0"];
       1 [label="2"];
       0 -- 1 [label=<+<sub>2</sub>>];
       2 [label="0" shape="box"];
       1 -- 2 [label=<z<sub>0</sub>>];
       3 [label="0" shape="box"];
       2 -- 3 [label=<*>];
       4 [label="($x ≡ z) ; ($y ≡ z)"];
       3 -- 4 [label=<*>];
       5 [label="0" shape="box"];
       0 -- 5 [label=<fst<sub>2</sub>>];
       6 [label="0" shape="box"];
       5 -- 6 [label=<*>];
       7 [label="0" shape="box"];
       6 -- 7 [label=<*>];
       8 [label="$p ≡ pair $f $s"];
       7 -- 8 [label=<*>];
       9 [label="0" shape="box"];
       0 -- 9 [label=<snd<sub>2</sub>>];
       10 [label="0" shape="box"];
       9 -- 10 [label=<*>];
       11 [label="0" shape="box"];
       10 -- 11 [label=<*>];
       12 [label="$p ≡ pair $f $s"];
       11 -- 12 [label=<*>];
       13 [label="2"];
       0 -- 13 [label=<Πv0>];
       14 [label="0" shape="box"];
       13 -- 14 [label=<El<sub>1</sub>>];
       15 [label="0" shape="box"];
       14 -- 15 [label=<*>];
       16 [label=< ⊆ FV(1)> shape="diamond"];
       15 -- 16 [label=<*>];
       17 [label="0" shape="box"];
       16 -- 17 [label=<✓>];
       18 [label="($b ≡ El $eb) ; (($a ≡ El $ea) ; ($c ≡ arrow $ea $eb))"];
       17 -- 18 [label=<*>];
       19 [label=<!>];
       16 -- 19 [label="!"];
}
